File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/reporting/DisplayTimer.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.reporting;

import android.app.Activity;

import com.urbanairship.UALog;
import com.urbanairship.Predicate;
import com.urbanairship.app.ActivityListener;
import com.urbanairship.app.ActivityMonitor;
import com.urbanairship.app.FilteredActivityListener;
import com.urbanairship.app.SimpleActivityListener;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;

/**
 * Tracks the amount of time displayed for the given {@code LifecycleOwner} or
 * {@code ActivityMonitor} and Activity predicate.
 * <p>
 * This class does not handle persisting timer state, which must be saved and restored from {@code savedInstanceState}
 * or elsewhere in order for tracking across pauses/resumes to remain accurate.
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class DisplayTimer {
    private long resumeTime = 0;
    private long displayTime = 0;

    /** Constructs a {@code DisplayTimer} with the default initial time of zero. */
    public DisplayTimer(@NonNull LifecycleOwner lifecycleOwner) {
        this(lifecycleOwner, 0);
    }

    /** Constructs a {@code DisplayTimer} with initial state set from the supplied {@code restoredDisplayTime}. */
    public DisplayTimer(@NonNull LifecycleOwner lifecycleOwner, long restoredDisplayTime) {
        if (restoredDisplayTime > 0) {
            displayTime = restoredDisplayTime;
        }

        lifecycleOwner.getLifecycle().addObserver(new LifecycleListener(this));
    }

    /** Constructs a {@code DisplayTimer} with the default initial time of zero. */
    public DisplayTimer(@NonNull ActivityMonitor activityMonitor) {
        this(activityMonitor, null, 0);
    }

    /** Constructs a {@code DisplayTimer} with initial state set from the supplied {@code restoredDisplayTime}. */
    public DisplayTimer(
        @NonNull ActivityMonitor activityMonitor,
        @Nullable Predicate<Activity> activityPredicate,
        long restoredDisplayTime
    ) {
        if (restoredDisplayTime > 0) {
            displayTime = restoredDisplayTime;
        }
        Predicate<Activity> predicate =
            activityPredicate != null ? activityPredicate : (activity) -> true;
        ActivityListener activityListener =
                new FilteredActivityListener(new DisplayActivityListener(this), predicate);

        activityMonitor.addActivityListener(activityListener);
    }


    /** Returns the current displayed time. */
    public long getTime() {
        long time = displayTime;
        if (resumeTime > 0) {
            time += System.currentTimeMillis() - resumeTime;
        }
        return time;
    }

    public void onResume() {
        resumeTime = System.currentTimeMillis();

    }

    public void onPause() {
        displayTime += System.currentTimeMillis() - resumeTime;
        resumeTime = 0;
    }

    private static final class DisplayActivityListener extends SimpleActivityListener {
        private final WeakReference<DisplayTimer> weakTimer;

        public DisplayActivityListener(DisplayTimer timer) {
            weakTimer = new WeakReference<>(timer);
        }

        @Override
        public void onActivityPaused(@NonNull Activity activity) {
            DisplayTimer timer = weakTimer.get();
            if (timer != null) {
                timer.onPause();
            } else {
                UALog.w("DisplayTimer ref was null!");
            }
        }

        @Override
        public void onActivityResumed(@NonNull Activity activity) {
            DisplayTimer timer = weakTimer.get();
            if (timer != null) {
                timer.onResume();
            } else {
                UALog.w("DisplayTimer ref was null!");
            }
        }
    }

    private static final class LifecycleListener implements DefaultLifecycleObserver {
        private final WeakReference<DisplayTimer> weakTimer;

        public LifecycleListener(DisplayTimer timer) {
            weakTimer = new WeakReference<>(timer);
        }

        @Override
        public void onResume(@NonNull LifecycleOwner owner) {
            DisplayTimer timer = weakTimer.get();
            if (timer != null) {
                timer.onResume();
            } else {
                UALog.w("DisplayTimer ref was null!");
            }
        }

        @Override
        public void onPause(@NonNull LifecycleOwner owner) {
            DisplayTimer timer = weakTimer.get();
            if (timer != null) {
                timer.onPause();
            } else {
                UALog.w("DisplayTimer ref was null!");
            }
        }

        @Override
        public void onDestroy(@NonNull LifecycleOwner owner) {
            owner.getLifecycle().removeObserver(this);
        }
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/reporting/DisplayTimer.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.reporting

import android.app.Activity
import androidx.annotation.RestrictTo
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.urbanairship.Predicate
import com.urbanairship.UALog
import com.urbanairship.app.ActivityMonitor
import com.urbanairship.app.FilteredActivityListener
import com.urbanairship.app.SimpleActivityListener
import com.urbanairship.util.Clock
import com.urbanairship.util.minus
import java.lang.ref.WeakReference
import java.time.Instant
import kotlin.time.Duration

/**
 * Tracks the amount of time displayed for the given `LifecycleOwner` or
 * `ActivityMonitor` and Activity predicate.
 *
 *
 * This class does not handle persisting timer state, which must be saved and restored from `savedInstanceState`
 * or elsewhere in order for tracking across pauses/resumes to remain accurate.
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class DisplayTimer private constructor(
    private val clock: Clock,
    restoredDisplayTime: Duration
) {

    /** The instant of the last [onResume], or `null` while paused. */
    private var resumeTime: Instant? = null

    /** Time accrued across completed resume/pause spans. */
    private var displayTime: Duration = restoredDisplayTime.coerceAtLeast(Duration.ZERO)

    public constructor(
        lifecycleOwner: LifecycleOwner,
        restoredDisplayTime: Duration = Duration.ZERO,
        clock: Clock = Clock.DEFAULT_CLOCK
    ) : this(clock, restoredDisplayTime) {
        lifecycleOwner.lifecycle.addObserver(LifecycleListener(this))
    }

    public constructor(
        activityMonitor: ActivityMonitor,
        activityPredicate: Predicate<Activity>? = null,
        restoredDisplayTime: Duration = Duration.ZERO,
        clock: Clock = Clock.DEFAULT_CLOCK
    ) : this(clock, restoredDisplayTime) {
        val predicate = activityPredicate ?: Predicate { activity: Activity -> true }
        val activityListener = FilteredActivityListener(
            listener = DisplayActivityListener(this),
            filter = predicate
        )

        activityMonitor.addActivityListener(activityListener)
    }

    public val time: Duration
        /** Returns the current displayed time.  */
        get() = resumeTime?.let { displayTime + (clock.now() - it) } ?: displayTime

    public fun onResume() {
        resumeTime = clock.now()
    }

    public fun onPause() {
        // Only accrue time if we were actually resumed. Previously `resumeTime` defaulted to 0,
        // so an unpaired onPause added the entire epoch-to-now span to the display time.
        resumeTime?.let { displayTime += clock.now() - it }
        resumeTime = null
    }

    private class DisplayActivityListener(timer: DisplayTimer): SimpleActivityListener() {

        private val weakTimer = WeakReference(timer)

        override fun onActivityPaused(activity: Activity) {
            val timer = weakTimer.get()?.apply { onPause() }
            if (timer == null) {
                UALog.w("DisplayTimer ref was null!")
            }
        }

        override fun onActivityResumed(activity: Activity) {
            val timer = weakTimer.get()?.apply { onResume() }
            if (timer == null) {
                UALog.w("DisplayTimer ref was null!")
            }
        }
    }

    private class LifecycleListener(timer: DisplayTimer): DefaultLifecycleObserver {

        private val weakTimer = WeakReference(timer)

        override fun onResume(owner: LifecycleOwner) {
            val timer = weakTimer.get()?.apply { onResume() }
            if (timer == null) {
                UALog.w("DisplayTimer ref was null!")
            }
        }

        override fun onPause(owner: LifecycleOwner) {
            val timer = weakTimer.get()?.apply { onPause() }
            if (timer == null) {
                UALog.w("DisplayTimer ref was null!")
            }
        }

        override fun onDestroy(owner: LifecycleOwner) {
            owner.lifecycle.removeObserver(this)
        }
    }
}
